### ISSUE TITLE
Unexpected response when container is near memory limit

### ISSUE THREAD

**Issue by lazorfuzz** [StackExchange/StackExchange.Redis#2804]
We're seeing the same issue posted here:
https://github.com/StackExchange/StackExchange.Redis/issues/2577

```
StackExchange.Redis.RedisConnectionException: Unexpected response to HMGET: Integer: 1
```
Triggered when calling `HashGetAsync` and when the app is running OOM. App is running on k8s in a container with a mem limit.

StackExchange.Redis version: `2.7.17`

Difficult to reproduce, it seems to happen when the application is close to the container mem limit. We see this happen alongside System.OutOfMemoryExceptions. 

Additionally, (and much scarier) behavior is we're also occasionally getting the wrong value on GETs in this state. E.g. We check the cache for `key:123`, but the data we get back is for `key:234`.

**Comment by mgravell** [comment]
Hmm, that is weird. It sounds like a protocol desync has occurred. This is very bad, and I plan to overhaul the IO core soon (prototype is active, etc). However, you could enable "high integrity mode" (see the configuration page, https://stackexchange.github.io/StackExchange.Redis/Configuration.html) - this will both confirm whether desync is the issue, and protect you from incorrect data fetched and all the consequences thereof. If it *does* report a desync, please let us know!

**Comment by mgravell** [comment]
Note that high integrity mode has a slight performance impact, since it adds an additional (tiny) command per operation; in almost all cases, this is effectively invisible and is not a concern.

**Comment by lazorfuzz** [comment]
Yep Linux/containerd. Base image is the standard Ubuntu distro from `mcr.microsoft.com/dotnet/aspnet:8.0`

**Comment by mgravell** [comment]
Thanks for the report. StackExchange.Redis **3.0** ships a substantial rewrite of the low-level IO core — a new RESPite-based stack that replaces `Pipelines.Sockets.Unofficial` and the old backlog/write loop — and we're confident this issue is addressed by that work.

Closing as resolved in 3.0. If you can still reproduce on 3.0.x, please comment and we'll reopen and investigate with fresh diagnostics.
